libusb/os/windows_winusb.c

```diff
@@ -1585,42 +1585,59 @@ static int winusb_get_device_list(struct libusb_context *ctx, struct discovered_
 					s = pRegQueryValueExA(key, "DeviceInterfaceGUID", NULL, &reg_type,
 						(LPBYTE)guid_string, &size);
 				pRegCloseKey(key);
-				if ((s == ERROR_SUCCESS) &&
-				    (((reg_type == REG_SZ) && (size >= (sizeof(guid_string) - sizeof(uint16_t))) && (size <= sizeof(guid_string))) ||
-				     ((reg_type == REG_MULTI_SZ) && (size >= sizeof(guid_string)) && (size <= (sizeof(guid_string) - sizeof(uint16_t)))))) {
-					if (nb_guids == guid_size) {
-						new_guid_list = realloc((void *)guid_list, (guid_size + GUID_SIZE_STEP) * sizeof(void *));
-						if (new_guid_list == NULL) {
-							usbi_err(ctx, "failed to realloc guid list");
-							LOOP_BREAK(LIBUSB_ERROR_NO_MEM);
-						}
-						guid_list = new_guid_list;
-						guid_size += GUID_SIZE_STEP;
-					}
-					if_guid = malloc(sizeof(*if_guid));
-					if (if_guid == NULL) {
-						usbi_err(ctx, "failed to alloc if_guid");
-						LOOP_BREAK(LIBUSB_ERROR_NO_MEM);
-					}
-					if (!string_to_guid(guid_string, if_guid)) {
-						usbi_warn(ctx, "device '%s' has malformed DeviceInterfaceGUID string '%s', skipping", dev_id, guid_string);
-						free(if_guid);
-					} else {
-						// Check if we've already seen this GUID
-						for (j = EXT_PASS; j < nb_guids; j++) {
-							if (memcmp(guid_list[j], if_guid, sizeof(*if_guid)) == 0)
-								break;
-						}
-						if (j == nb_guids) {
-							usbi_dbg("extra GUID: %s", guid_string);
-							guid_list[nb_guids++] = if_guid;
-						} else {
-							// Duplicate, ignore
-							free(if_guid);
-						}
-					}
-				} else if (s == ERROR_SUCCESS) {
-					usbi_warn(ctx, "unexpected type/size of DeviceInterfaceGUID for '%s'", dev_id);
+				if ( s == ERROR_SUCCESS)
+				{
+                    /// https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexa#remarks
+                    /// - REG_SZ: "string may not have been stored with the proper terminating null characters."
+                    /// - As REG_SZ but also "Note that REG_MULTI_SZ strings could have two terminating null characters"
+                    unsigned int minREG_SZ = sizeof(guid_string) - sizeof(uint16_t);
+                    unsigned int maxREG_SZ = sizeof(guid_string);   
+                    
+                    /// As the reading code does not read enough for REG_MULTI_SZ we can handle it as REG_SZ
+				    if ( ((reg_type == REG_SZ) || (reg_type == REG_MULTI_SZ)) 
+                        && (size >= minREG_SZ) 
+                        && (size <= maxREG_SZ) ) 
+                    {
```

Comment:
Did you change anything below here except whitespace and the todo comment? It is hard to tell in the web interface.

Reply:
I pushed with the indentation fixed which is easier to see. The indentation wasn't correct in the orignal code so to see just what I changed you can check 'ignore whitespace changes' for the diff.
1. Simplified REG_SZ checking logic
2. Removed duplicated check for  `s == ERROR_SUCCESS`

Reply:
This is the diff I used to check the changes: https://github.com/libusb/libusb/pull/896/files?diff=split&w=1

Reply:
Ah thanks, I didn't know I could filter whitespace changes so easily here.

Alternatively, deal with other errors up front, and avoid any indentation changes:
```
+                               if (s != ERROR_SUCCESS) {
+                                       usbi_warn(ctx, "unexpected error from pRegQueryValueExA for '%s'", dev_id);
+                                       break;
+                               }
```
 
Now, to understand the whole change better, what is it what we get from pRegQueryValueExA()? Isn't it a string with the 38 characters (braces or parenthesis) format: {12345678-1234-1234-1234-123456789abc} ?

Written out (not binary encoded) there are 32 hex digits in a GUID plus the dashes and braces, so 38 string characters. I reread your bug report and it talks about 19 unicode characters? Does that make sense at all? If the GUID string was encoded in unicode it would be 76 bytes... I think unicode is not relevant here, and the double zero at the end would not be a unicode \0 but the double zero termination in a MS REG_MULTI_SZ.

38 bytes if it is a GUID without trailing zero and 39 if it is a GUID with one trailing zero (the only case supported up to now?) and 40 if it is a GUID with double zeroes (from REG_MULTI_SZ)? Should we accept 40 for a REG_SZ? Shouldn't we also check that these extra bytes are effectively zeroes? Or is it redundant because of string_to_guid()?

Reply:
What is the status of this? Should it be taken as-is or are additional changes needed?

Reply:
If my above comment is correct, and the string_to_guid() takes the final check, I suggest the simpler change in pull request #927. I left Craig as author, but his e-mail address for sign-off should probably be changed, I only found the github-scrambled one. The difference is much a question of taste, but at least replacing sizeof(uint16_t) with 2*sizeof(char) is important for readability and correctness.